core-to-plc/src/Language/Plutus/CoreToPLC.hs

```diff
@@ -301,8 +301,20 @@ PLC is strict, but users *do* expect that, e.g. they can write an if expression
 lazy. This only *matters* because we have 'error', so it's important that 'if false error else ...'
 does not evaluate to 'error'!
 
-More generally, we compile case expressions (of which an if expression is one) lazily, i.e. we add
+More generally, we can compile case expressions (of which an if expression is one) lazily, i.e. we add
 a unit argument and apply it at the end.
+
+However, we apply an important optimization: we only need to do this if it is not the case that
+all the case expressions are branches. In the common case they *will* be, so this gives us
+significantly better codegen a lot of the time.
+
+The check we do is:
+- Alternatives with arguments will be turned into lambdas by us, so will be values.
```

Comment:
This makes the whole approach sound, so I think it needs some stress. Because otherwise we could sometimes generate eta-contracted functions (i.e. not lambdas) in a well-typed typed and that would break the lazy semantics of `case`.

Reply:
I don't think we will ever do that, but sure.

Reply:
Actually, I don't see how we're ever going to translate a case branch with arguments as anything other than a lambda - in theory the user could write a function with the right type, but they have no way to use it in the match afaik.

Reply:
I do see your point, I just don't see why we would do that, and I think the current description makes it clear enough.